Mark stale request as dirty

ferpesin opened the pull request.

jfarre90 commented:
Performed A-B-A testing with and without the useQuery changes to confirm the test fails without them.

As an explanation on why we need to wait multiple "flushes" in the test:
The use of flushSync  inside useQuery is important --> it forces React to synchronously update the DOM after both the cache update and the state dispatch. This means each patch operation triggers:
- A cache update
- A state update to 'success'
- Potentially another refetch due to realtime: 'refetch'

When a patch comes in with realtime: 'refetch', this effect triggers a new fetch. If two patches come in quickly (like in the test/scenario we are trying to fix), we would have:
- First patch triggers refetch
- Second patch comes in before first refetch completes
- Second refetch starts
- Results could arrive in any order

Each refetch sets the state back to pending, which triggers another fetch cycle. Each update follows this cycle:
- Patch comes in
- Triggers a refetch
- Updates cache
- Updates state
- Component re-renders
- 
The test was failing before because we weren't giving enough time for this cycle to complete. The polling approach we took in the test works because it await the 10ms each loop until the component re-rendering has reached its final state to settle after all the updates

KidkArolis commented on lib/useQuery.js:
```diff
@@ -110,18 +117,26 @@ export function useQuery(serviceName, options = {}, queryHookOptions = {}) {
         }
       })
+      .finally(() => {
+        initialisedRef.current = true
```
@ferpesin @jfarre90 We set this to true, but never back to false - that doesn't seem right?

jfarre90 replied:
I believe the intention of this was to get around some bug on the first render/fetch, for which we did not want to trigger the "reset" and cause a refetch. @ferpesin correct me if I missed any more info.

The intention is that it behaves as a "first load completed". So the intention of only setting it to true once is intentional. Once the first load is completed, we want this to always stay as true.

KidkArolis commented:
Could you get this rebased please!

jfarre90 committed:
Refactor test to await for the correct state updates

jfarre90 commented:
> Could you get this rebased please!

@KidkArolis Done!

KidkArolis committed:
Replace the initialisedRef.current with previousQueryIdRef check

To me - initialisedRef.current looked like it would cause a bug
in case the query changes _during_ the first fetch, it would never
refetch because we would have skipped over that case. With the
current approach we always call refetch if queryId changes, but
we no longer call refetch on the initial render

KidkArolis committed:
Add an additional test to check .refetch() works during ongoing request

KidkArolis commented on lib/useQuery.js:
```diff
@@ -110,15 +116,19 @@ export function useQuery(serviceName, options = {}, queryHookOptions = {}) {
     })
       .then(res => {
-        if (reqRef === requestRef.current) {
+        if (state.dirty) {
```
It is not correct / valid to read state.dirty in a promise callback because this state variable is a stale closure - this will not reflect the state as of now, but as it was when the effect was first triggered.
So while the tests seem to be passing, I don't think this is 100% correct. 🤔 But also - really don't know how to correct it - the new branch where state is managed outside React is much easier to reason about.. should we get that over the finish line??

KidkArolis commented on lib/useQuery.js:
```diff
@@ -137,16 +147,24 @@ export function useQuery(serviceName, options = {}, queryHookOptions = {}) {
+    previousQueryIdRef.current = queryId
+  }, [dispatch, queryId, isCacheSufficient])
+
+  const refetch = useCallback(() => dispatch({ type: 'reset' }), [dispatch])
```
I removed the introduction of initialisedRef, because I do not think it was modelling the behaviour correctly. We only marked initialisedRef to true after the fetch promise completed, but the query could have changed in the meantime and we would never refetch - that could introduce subtle bugs.

With the previousQueryIdRef initialised to null, we solve the same issue of avoiding fetching on initial render - but we do initiate refetch when query changes.

KidkArolis committed:
Remove unused condition from the effect

KidkArolis commented:
This dirty bug is fixed now in the 0.20.x series on npm and live in our app!
